### ISSUE TITLE
Ineffective loop with actions

### ISSUE THREAD

**Issue by slapin** [reattiva/Urho3D-Blender#60]
This is generally more an enhancement, but that might mean having minutes of export time vs days...

Well, the animation export script single-steps whole blender bones*frames times,
which leads to tiny scene complication increases (i.e. unapplied modifers) increase export times
to huge extents (like from minutes to days). I think loop should be changed like this:

1. Find-out maximum action length. Loop through steps from 0 to max.
2. For each mesh for each action - if we have current frame step in action then
add to track, passing through all meshes and all actions.
3. write-out animations.
This way it should become much more quicker even for best-case scenarios with very simple
scenes. For me adding single decimate modifier leads from 3 min export time to 28 hour export time.
This is really surprised me...
This modifier don't have to be in exported object, just for something on the scene.
So loop rearrangement should do very good improvement with performance.

**Comment by reattiva** [comment]
Does checking "Apply modified" help?
The way animations are exported is the only one I know, so I need some working code from you...
By the way, this work is a bit too old for me.

**Comment by slapin** [comment]
Well, "Apply modified" doesn't affect both stepping time and final export
time.
The more unapplied modifiers there the slower stepping is. The only
possible optimization is
to do less stepping.

**Comment by reattiva** [comment]
Yes, I was drunk, if anything "Apply modifiers" will increase the exporting time...

**Comment by Mike3D** [comment]
I think the core issue lies in the fact that scene.frame_set() updates the whole scene (any object), including lights, empties, meshes, modifiers...

As there doesn't seem to be alternatives to it, you could try to temporarilly put your armature in a different scene (as it can belong to multiple scenes) and call frame_set() for this new scene. That way only the armature will be updated.
We could then spare a few ms by simply swaping bone <-> frame loop.

Assignment:
- at the beginning of DecomposeActions():
    - create a temporary scene : 'myTempScene'
    - make your armature belong to myTempScene
- frame decomposition : call myTempScene.frame_set() instead of scene.frame_set()
- at the end of DecomposeActions() : delete myTempScene

**Comment by Mike3D** [comment]
Code (not tested):
```python
# Work with armature in a temporary scene
myTempScene = bpy.data.scenes.new('myTempScene')
myTempScene.objects.link(armatureObj)
...
myTempScene.frame_set()
...
# Remove temporary scene
myTempScene.objects.unlink(armatureObj)
bpy.data.scenes.remove(myTempScene)

**Comment by Mike3D** [comment]
Quick test using a 20 frames anim on 38 bones, with a SubSurf modifier (313k polys):
- current state : 277s (almost 5mins!) :scream:
- with temporary scene : 2s :stuck_out_tongue_winking_eye:

Could MakeHuman users (@vivienneanthony, @slapin, ...) please confirm the test before considering PR

**Comment by reattiva** [comment]
Does the "read actions by f-curves" option help? Or is it unusable by most cases?

**Comment by Mike3D** [comment]
The "read actions by f-curves" only applies to actions, which are already efficiently decomposed.
Every other kind of animation pass through the frame_set() update, that's where the issue lies.

I've done a quick test swapping bones and frames in the for loop and got additional 0.6 seconds gain (from 2.0 to 1.4). This improvement applies to any kind of animation.

I'll do an ultimate test making the temporary scene active (so that no rendering occurs at all), as I've noticed that without the SubSurf modifier the export time drops to 0.14 seconds!

**Comment by slapin** [comment]
Ah, thank you so much! Will test in a day.

**Comment by Mike3D** [comment]
@reattiva, exporting timeline animation as action using 'Fcurves' option works great. It is slightly slower than using a temporary scene, maybe due to individual keyframes interpolation (individual keyframes don't need interpolation).

**Comment by reattiva** [comment]
Good! If the temp scene works better, then I'm all for it and we can scratch the f-curves and their limitations.

**Comment by Mike3D** [comment]
Thanks. I'll do some more tests (more bones, more frames) to confirm the benefits and test some other ideas (work on a copy of the armature...).
I'll also dig into the FCurves API to check for further improvements.
